### PR TITLE
Fix spec and shape inference for Unsqueeze op

### PR DESCRIPTION
**Pull request by hariharans29**
1) Unsqueeze op is unclear on how it deals with values in `axes` and what are the acceptable values for `axes`

2) Fix shape inference logic. In current state, it cannot handle negative axes properly (the way it is intended to)

### CODE REVIEW

**Comment by hariharans29** [comment]
CC: @wschin @gramalingam

File: onnx/test/shape_inference_test.py [hunk 1]
```diff
@@ -455,13 +455,27 @@ def test_squeeze(self):  # type: () -> None
...
         self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.FLOAT, (1, 1, 3, 1, 2, 1))])
 
+    def test_unsqueeze_2(self):  # type: () -> None
+        graph = self._make_graph(
+            [('x', TensorProto.FLOAT, (3, 4, 5))],
+            [make_node('Unsqueeze', 'x', 'y', axes=[0, 4])],
+            [])
+        self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.FLOAT, (1, 3, 4, 5, 1))])
+
+    def test_unsqueeze_negative_axes(self):  # type: () -> None
```
Comment (hariharans29) [inline onnx/test/shape_inference_test.py]: won't work with existing logic.

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -1301,11 +1301,17 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-  Given a tensor such that tensor with shape [3, 4, 5], then
-  Unsqueeze(tensor, axes=[0, 4]) has shape [1, 3, 4, 5, 1]
+Insert single-dimensional entries to the shape of an input tensor (`data`).
+Takes one required argument `axes` - which contains a list of dimension indices and this operator will insert a dimension of value `1` into the corresponding index of the output tensor (`expanded`).
+
+For example:
+  Given an input tensor (`data`) of shape [3, 4, 5], then
+  Unsqueeze(data, axes=[0, 4]) outputs a tensor (`expanded`) containing same data as `data` but with shape [1, 3, 4, 5, 1].
+
+The attribute `axes` should not contain any duplicate entries. It is an error if it contains duplicates.
```
Comment (spandantiwari) [inline onnx/defs/tensor/defs.cc]: +1 for mentioning about duplicate entries.

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -1301,11 +1301,17 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-Takes one required argument `axes`, a list of dimensions that will be inserted.
-Dimension indices in `axes` are as seen in the output tensor. For example:
-  Given a tensor such that tensor with shape [3, 4, 5], then
-  Unsqueeze(tensor, axes=[0, 4]) has shape [1, 3, 4, 5, 1]
+Insert single-dimensional entries to the shape of an input tensor (`data`).
+Takes one required argument `axes` - which contains a list of dimension indices and this operator will insert a dimension of value `1` into the corresponding index of the output tensor (`expanded`).
+
+For example:
+  Given an input tensor (`data`) of shape [3, 4, 5], then
+  Unsqueeze(data, axes=[0, 4]) outputs a tensor (`expanded`) containing same data as `data` but with shape [1, 3, 4, 5, 1].
```
Comment (spandantiwari) [inline onnx/defs/tensor/defs.cc]: As discussed, we should consider adding a note clarifying that the order of axes specified in the `axes` attribute does not matter, and that they should be sorted before being applied.

**Comment by wschin** [comment]
Do we have tests for unsorted cases?

File: onnx/defs/tensor/defs.cc (hunk 3 above)
Comment (hariharans29) [reply onnx/defs/tensor/defs.cc]: I added this line - 

The order of values in `axes` does not matter and can come in any order.

I think this is good for a high level spec. Details like the operator will sort them after correcting negative values are superfluous, I feel. What do you think ?

**Comment by hariharans29** [comment]
> Do we have tests for unsorted cases?

I will add shape inference test for now.

**Comment by hariharans29** [comment]
@spandantiwari @wschin - added backend test for unsorted axes too

File: onnx/backend/test/case/node/unsqueeze.py [hunk 4]
```diff
@@ -61,6 +61,23 @@ def export_unsqueeze_three_axes():  # type: () -> None
...
 
+    @staticmethod
+    def export_unsqueeze_unsorted_axes():  # type: () -> None
+        x = np.random.randn(3, 4, 5).astype(np.float32)
+
+        node = onnx.helper.make_node(
+            'Unsqueeze',
+            inputs=['x'],
+            outputs=['y'],
+            axes=[5, 4, 2],
```
Comment (hariharans29) [inline onnx/backend/test/case/node/unsqueeze.py]: axes values in jumbled order

**Pull request merged**